### PR TITLE
Displacements Now Have Lightmaps

### PR DESCRIPTION
**Pull request by Fronkln**
Ported over from Source to Unity, tested and confirmed. Biggest contribution yet!
image

image

### CODE REVIEW

**Comment by Fronkln** [comment]
There seems to be a super rare interesting bug with this. Some of the displacements are placed with wrong offset

image

Results in this:
image

Should probably not be anything difficult to find out/fix? i sure as hell don't have any clue though

**Comment by Fronkln** [comment]
Not entirely sure but the UV's might need to be flipped by 90 degrees for displacements

**Comment by lewa-j** [comment]
Already solved this thing in a different project. Look here for reference https://github.com/lewa-j/vbsp-gles/blob/master/src/bsp_loader.cpp#L1014

**Comment by Fronkln** [comment]
Alright, i'll take a look

**Comment by Fronkln** [comment]
I believe its easier and better to just rewrite displacement code accordingly to the bsp loader you linked, i'll follow this up with another pull request if i manage to get it working

**Comment by Fronkln** [comment]
Managed to fix it commit tomorrow

**Comment by Fronkln** [comment]
SDK 2013 (Build2001 mod city_test03) (rubble displacements)
image

Half-Life 2 Beta (quarry_town02) (entire road is a displacement)
image

File: Assets/Code/Read/SourceBSPLoader.cs [hunk 1]
```diff
@@ -892,6 +892,12 @@ void GenerateDispFaceObject (int index, int model)
 			mf.sharedMesh.uv = f.uv;
 			mf.sharedMesh.colors32 = f.cols;
 
+           	        if(uSrcSettings.Inst.lightmaps)
+			{
+               			mr.lightmapIndex = curLightmap;
```
Comment (lewa-j) [inline Assets/Code/Read/SourceBSPLoader.cs:897]: Here

File: Assets/Code/Read/SourceBSPLoader.cs [hunk 2]
```diff
@@ -1037,13 +1043,35 @@ surface BuildDispFace (int faceIndex, int model, short dispinfoId)
...
 
 			bspface curFace = map.facesLump[faceIndex];
 
 			bsptexinfo curTexInfo = map.texinfosLump[curFace.texinfo];
 			bsptexdata curTexData = map.texdataLump[curTexInfo.texdata];
+		        
+			int tiFlags = curTexInfo.flags;
+
+        	        int lightmapW = curFace.LightmapTextureSizeInLuxels[0] + 1;
+                        int lightmapH = curFace.LightmapTextureSizeInLuxels[1] + 1;
```
Comment (lewa-j) [inline Assets/Code/Read/SourceBSPLoader.cs:1058]: here

File: Assets/Code/Read/SourceBSPLoader.cs [hunk 3]
```diff
@@ -1037,13 +1043,35 @@ surface BuildDispFace (int faceIndex, int model, short dispinfoId)
...
+
+                        int lmx = 0, lmy = 0;
+
+          	        if (uSrcSettings.Inst.lightmaps && (tiFlags & SourceBSPStructs.SURF_NOLIGHT) == 0)
+                        {
```
Comment (lewa-j) [inline Assets/Code/Read/SourceBSPLoader.cs:1063]: this entire block is a mess

File: Assets/Code/Read/SourceBSPLoader.cs [hunk 4]
```diff
@@ -1108,7 +1136,24 @@ surface BuildDispFace (int faceIndex, int model, short dispinfoId)
...
+           	        origLMUV[1] = new Vector2(0.5f, lightmapH - 0.5f);
+           	        origLMUV[2] = new Vector2(lightmapW - 0.5f, lightmapH - 0.5f);
+           	        origLMUV[3] = new Vector2(lightmapW - 0.5f, 0.5f);
+			
+			for (int j = 0; j < 4; j++)
+             		        origLMUV[j] = (origLMUV[j] + new Vector2(lmx, lmy)) / (float)BLOCK_SIZE;
+			
+		        Vector2[] dXlmuv = new Vector2[2]
+       		        {
+              		       (origLMUV[1] - origLMUV[0]) * invHeight,
```
Comment (lewa-j) [inline Assets/Code/Read/SourceBSPLoader.cs]: here

File: Assets/Code/Read/SourceBSPLoader.cs [hunk 5]
```diff
@@ -1118,6 +1163,14 @@ surface BuildDispFace (int faceIndex, int model, short dispinfoId)
 
 				leftRightSeg = rightEnd - leftEnd;
 				leftRightStep = leftRightSeg * subdivideScale;
+				
+			        Vector2[] dYlmuv = new Vector2[2]
+               		        {
```
Comment (lewa-j) [inline Assets/Code/Read/SourceBSPLoader.cs]: here

**Review by lewa-j — CHANGES_REQUESTED** [review]

File: Assets/Code/Read/SourceBSPLoader.cs [hunk 6]
```diff
@@ -1134,8 +1187,8 @@ surface BuildDispFace (int faceIndex, int model, short dispinfoId)
 
 					float tU = Vector3.Dot (flatVertex, curTexInfo.texvecs) + (curTexInfo.texoffs);
 					float tV = Vector3.Dot (flatVertex, curTexInfo.texvect) + (curTexInfo.texofft);
-					UVs.Add (new Vector2 (tU * scaleU, tV * scaleV));
-
+					UVs.Add (new Vector2 (tU * scaleU, tV * scaleV));
```
Comment (lewa-j) [inline Assets/Code/Read/SourceBSPLoader.cs:1188]: trailing spaces here

**Comment by lewa-j** [comment]
Fix identation and trailing spaces please.

**Comment by Fronkln** [comment]
I hate it when Github does this, i'll take a quick look

File: Assets/Code/Read/SourceBSPLoader.cs [hunk 7]
```diff
@@ -1051,26 +1051,25 @@ surface BuildDispFace (int faceIndex, int model, short dispinfoId)
...
-		        
+
+            int tiFlags = curTexInfo.flags;
 			int tiFlags = curTexInfo.flags;
 
             int lightmapW = curFace.LightmapTextureSizeInLuxels[0] + 1;
             int lightmapH = curFace.LightmapTextureSizeInLuxels[1] + 1;
 
-            int lmx = 0, lmy = 0;
+            int lmx = 0; lmy = 0;
```
Comment (lewa-j) [inline Assets/Code/Read/SourceBSPLoader.cs]: Now it won't compile

File: Assets/Code/Read/SourceBSPLoader.cs [hunk 8]
```diff
@@ -1051,26 +1051,25 @@ surface BuildDispFace (int faceIndex, int model, short dispinfoId)
...
 
-            if (uSrcSettings.Inst.lightmaps && (tiFlags & SourceBSPStructs.SURF_NOLIGHT) == 0)
+            if (usrcSettings.Inst.lightmaps && (tiFlags & SourceBSPStructs.SURF_NOLIGHT) == 0)
```
Comment (lewa-j) [inline Assets/Code/Read/SourceBSPLoader.cs]: why?

**Comment by lewa-j** [comment]
I has nothing to do with github. I believe the cause is your ide. Now we have mixed up spaces and tabs.

**Comment by Fronkln** [comment]
Visual studio is playing some big games with me right now, despite the code appearing normal in vs, i'll reopen this when im sure its fixed

I'm really sorry im puttin you through this

**Comment by lewa-j** [comment]
Try to open files in something like notepad++. It can show you where spaces and tab characters are

**Comment by lewa-j** [comment]
And instead of closing and opening new PRs, and creating many unnecessary commits, you should learn how to force push commits.

File: Assets/Code/Read/SourceBSPLoader.cs (hunk 7 above)
Comment (antim0118) [reply Assets/Code/Read/SourceBSPLoader.cs]: wtf

**Pull request closed without merging**